src/dune_engine/scheduler.ml

```diff
@@ -759,45 +782,62 @@ let maybe_clear_screen ~config =
          ])
 
 let poll ?config ~once ~finally () =
-  let t = prepare ?config () in
-  let watcher = File_watcher.create () in
-  let block_waiting_for_changes () =
-    match Event.next () with
-    | Job_completed _ -> assert false
-    | Files_changed -> Continue
-    | Signal signal ->
-      got_signal signal;
-      Exit
-  in
-  let wait msg =
-    Console.Status_line.set_temporarily
-      (fun () ->
-        Some (Pp.seq msg (Pp.verbatim ", waiting for filesystem changes...")))
-      (fun () ->
-        let res = block_waiting_for_changes () in
-        maybe_clear_screen ~config;
-        res)
+  let t = prepare ?config ~polling:true () in
+  let finally () =
+    (* Re-create the throttle so that we don't keep old ivars from one run to
+       the next *)
+    t.job_throttle <- Fiber.Throttle.create (Fiber.Throttle.size t.job_throttle);
+    Fiber.return (finally ())
   in
+  let watcher = File_watcher.create () in
   let rec loop () =
-    let res = Run_once.run_and_cleanup t once in
-    finally ();
-    match res with
+    t.status <- Building;
+    let open Fiber.O in
+    let* res =
+      let on_error exn () =
+        match t.status with
+        | Building -> Report_error.report exn
+        | Cancelling_build -> ()
+        | Waiting_for_file_changes _ ->
+          (* We are inside a build, so we aren't waiting for a file change event *)
+          assert false
+      in
+      Fiber.fold_errors ~init:() ~on_error (fun () ->
+          Fiber.finalize ~finally once)
```

Comment:
Is it important to have `fold_errors` scope around `finalize`?
If we swap them, then the code is probably simpler to reason about because you don't need `Fiber.finalize` at all (and perhaps it makes error reporting happen earlier, not sure).

Reply:
Errors are already reported immediately by `with_error_handler` in `Run_once.run_and_cleanup`.

We need `fold_errors` to report errors from the finalizer.

Reply:
Ok, makes sense. Then wouldn't have `once` outside of `fold_errors` make the intention clearer?

Reply:
But we still need to catch errors from `once`. Without `fold_errors`, we will not whether to report `Success` or `Had Errors`.

Reply:
I ended up getting rid of the error handling of `finally`. If the finalizer raises, that's basically a fatal error and it's not safe to restart the build.

Reply:
Ok, seems fine.